src/levels.rs

```diff
@@ -1,5 +1,1053 @@
+                            // do nothing
+                        } else {
+                            num_skips += 1;
+                            // TODO: update stats of vlog
+
+                            iter.next();
+                            continue;
+                        }
+                    }
+                }
+
+                num_keys += 1;
+
+                if vs.meta & crate::value::VALUE_POINTER != 0 {
+                    vp.decode(&vs.value);
+                }
+
+                builder.add(&iter_key, &vs, vp.len);
+
+                iter.next();
+            } // Finish iteration.
+
+            if builder.is_empty() {
+                continue;
+            }
+
+            let file_id = self.reserve_file_id();
+
+            let table = if self.opts.in_memory {
+                Table::open_in_memory(builder.finish(), file_id, bopts)?
+            } else {
+                let filename = crate::table::new_filename(file_id, &self.opts.dir);
+                Table::create(&filename, builder.finish(), bopts)?
+            };
+
+            tables.push(table);
+        }
+
+        eprintln!(
+            "compactor {}, sub_compact took {} mills, produce {} tables, added {} keys, skipped {} keys",
+            compact_def.compactor_id,
+            std::time::Instant::now()
+                .duration_since(start_time)
+                .as_millis(),
+            tables.len(),
+            num_keys,
+            num_skips
+        );
+
+        Ok(tables)
+    }
+
+    /// Merges top tables and bot tables to form a list of new tables.
+    fn compact_build_tables(
+        self: &Arc<Self>,
+        level: usize,
+        compact_def: &CompactDef,
+        _pool: &yatp::ThreadPool<yatp::task::callback::TaskCell>,
+    ) -> Result<Vec<Table>> {
+        // TODO: this implementation is very very trivial
+
+        let mut valid = vec![];
+
+        for table in &compact_def.bot {
+            // TODO: Check valid.
+            valid.push(table.clone());
+        }
+
+        let valid = Arc::new(valid);
+
+        let make_iterator = move |compact_def: &CompactDef, valid: &[Table]| {
+            let mut iters = vec![];
+
+            if level == 0 {
+                for table in compact_def.top.iter().rev() {
+                    iters.push(TableIterators::from(
+                        table.new_iterator(crate::table::ITERATOR_NOCACHE),
+                    ));
+                }
+            } else if !compact_def.top.is_empty() {
+                assert_eq!(compact_def.top.len(), 1);
+                iters.push(TableIterators::from(
+                    compact_def.top[0].new_iterator(crate::table::ITERATOR_NOCACHE),
+                ));
+            }
+
+            // TODO: Use ConcatIterator.
+            for table in valid {
+                iters.push(TableIterators::from(
+                    table.new_iterator(crate::table::ITERATOR_NOCACHE),
+                ));
+            }
+            iters
```

Comment:
Confused by this `iters` it contains tables in bot & top, but when construct the MergeIterator in merge_iterator.rs L184 how to guarantee `is_left_small` is true since bot tables may overlap with top tables.

Reply:
We can refer to `fill_tables*` to know how to fill tables for `CompactDef`. The key function inside it is `LevelHandler::overlapping_tables`, that guarantees that tables of `bot` in `CompactDef` will never be smaller than those of `top`.